File: src/arboles/java/Tree.java
package arboles.java;

import java.util.LinkedList;
import java.util.Queue;

public class Tree {
    private Node root;

    public Tree() {
        root = null;
    }

    public void add(int value) {
        root = addRecursive(root, value);
    }

    private Node addRecursive(Node current, int value) {
        if (current == null) {
            return new Node(value);
        } else if (value < current.getValue()) {
            current.setLeft(addRecursive(current.getLeft(), value));
        } else if (value > current.getValue()) {
            current.setRight(addRecursive(current.getRight(), value));
        }

        return current;
    }

    public boolean containsNode(int value) {
        return containsNodeRecursive(root, value);
    }

    private boolean containsNodeRecursive(Node current, int value) {
        if (current == null) {
            return false;
        } else if (value == current.getValue()) {
            return true;
        }

        return value < current.getValue() ?
                containsNodeRecursive(current.getLeft(), value) :
                containsNodeRecursive(current.getRight(), value);
    }

    public void delete(int value) {
        root = deleteRecursive(root, value);
    }

    private Node deleteRecursive(Node current, int value) {
        if (current == null) {
            return null;
        } else if (value == current.getValue()) {
            if (current.getLeft() == null && current.getRight() == null) {
                return null;
            } else if (current.getRight() == null) {
                return current.getLeft();
            } else if (current.getLeft() == null) {
                return current.getRight();
            } else {
                int smallestValue = findSmallestValue(current.getRight());
                current.setValue(smallestValue);
                current.setRight(deleteRecursive(current.getRight(), smallestValue));
            }
        } else if (value < current.getValue()) {
            current.setLeft(deleteRecursive(current.getLeft(), value));
        } else {
            current.setRight(deleteRecursive(current.getRight(), value));
        }
        return current;
    }

    private int findSmallestValue(Node root) {
        return root.getLeft() == null ?
                root.getValue() :
                findSmallestValue(root.getLeft());
    }

    public void traverseInOrder(Node node) {
        if (node != null) {
            traverseInOrder(node.getLeft());
            System.out.print(" " + node.getValue());
            traverseInOrder(node.getRight());
        }
    }

    public void traversePreOrder(Node node) {
        if (node != null) {
            System.out.print(" " + node.getValue());
            traversePreOrder(node.getLeft());
            traversePreOrder(node.getRight());
        }
    }

    public void traversePostOrder(Node node) {
        if (node != null) {
            traversePostOrder(node.getLeft());
            traversePostOrder(node.getRight());
            System.out.print(" " + node.getValue());
        }
    }

    public void traverseLeverOrder() {
        if (root == null){
            return;
        }

        Queue<Node> nodes = new LinkedList<>();
        nodes.add(root);

        while (!nodes.isEmpty()) {
            Node node = nodes.remove();

            System.out.print(" " + node.getValue());

            if (node.getLeft() != null) {
                nodes.add(node.getLeft());
            } else if (node.getRight() != null) {
                nodes.add(node.getRight());
            }
        }
    }

    public Node getRoot() {
        return root;
    }
}

File: src/arboles/kotlin/Tree.kt
package arboles.kotlin

import java.util.*

class Node(var value: Int) {
    var right: Node? = null
    var left: Node? = null
}

class Tree {
    private var root: Node? = null

    fun add(value: Int) {
        root = addRecursive(root, value)
    }

    private fun addRecursive(current: Node?, value: Int): Node? {
        when {
            current == null -> return Node(value)
            value < current.value -> current.left = addRecursive(current.left, value)
            value > current.value -> current.right = addRecursive(current.right, value)
        }
        return current
    }

    fun containsNode(value: Int): Boolean {
        return containsNodeRecursive(root, value)
    }

    private fun containsNodeRecursive(current: Node?, value: Int): Boolean {
        return if (current == null) {
            false
        } else if (value == current.value) {
            true
        } else {
            if (value < current.value) {
                containsNodeRecursive(current.left, value)
            } else {
                containsNodeRecursive(current.right, value)
            }
        }
    }

    fun delete(value: Int) {
        root = deleteRecursive(root, value)
    }

    private fun deleteRecursive(current: Node?, value: Int): Node? {
        return when {
            current == null -> null
            value == current.value -> {
                if (current.left == null && current.right == null) {
                    null
                } else if (current.right == null) {
                    current.left
                } else if (current.left == null) {
                    current.right
                } else {
                    val smallestValue = findSmallestValue(current.right)
                    current.value = smallestValue
                    current.right = deleteRecursive(current.right, smallestValue)
                    current
                }
            }
            value < current.value -> {
                current.left = deleteRecursive(current.left, value)
                current
            }
            else -> {
                current.right = deleteRecursive(current.right, value)
                current
            }
        }
    }

    private fun findSmallestValue(root: Node?): Int {
        return if (root!!.left == null) root.value else findSmallestValue(root.left)
    }

    fun traverseInOrder(node: Node?) {
        if (node != null) {
            traverseInOrder(node.left)
            print(" ${node.value}")
            traverseInOrder(node.right)
        }
    }

    fun traversePreOrder(node: Node?) {
        if (node != null) {
            print(" ${node.value}")
            traversePreOrder(node.left)
            traversePreOrder(node.right)
        }
    }

    fun traversePostOrder(node: Node?) {
        if (node != null) {
            traversePostOrder(node.left)
            traversePostOrder(node.right)
            print(" ${node.value}")
        }
    }

    fun traverseLevelOrder() {
        if (root == null) {
            return
        }

        val nodes: Queue<Node> = LinkedList()
        nodes.add(root)

        while (!nodes.isEmpty()) {
            val node = nodes.remove()

            print(" ${node.value}")

            if (node.left != null) {
                nodes.add(node.left)
            } else if (node.right != null) {
                nodes.add(node.right)
            }
        }
    }

    fun getRoot(): Node? {
        return root
    }
}